### PR TITLE
build: cli tools package

### PR DESCRIPTION
**Pull request by karpov-kir**
An attempt to encapsulate some common or tricky build-related logic to a new package. Everything is now driven by `NODE_ENV`.

High-level idea:

- Introduce `@dddforum/cli` package that encapsulates commands to simplify scripting. For example, there is `prepareEnv` command:
  - Checks docker
  - Starts docker
  - Migrates DB
  - Build Prisma client
- The new package exposes commands via [an executable file `dfCli <command>`](https://github.com/stemmlerjs/ddd-forum-bp-first/pull/7/files#diff-6e2e2a1851648938b325ba84de634407a4e69a644ea61102df15ca4a8a7a9758R7)
  - This allows to use it simply `NODE_ENV=development dfCli prepareEnv -d ...`
- Introduce `ensureAndLoadEnv` helper that can be used in whatever place to load env variables driven by `NODE_ENV` e.g. when starting the backend, when running tests, etc. This helper loads the closest `.env.${NODE_ENV}` file.
- Use commands from `@dddforum/cli` in NPM scripts
- Control which env is used by `NODE_ENV` (for example check scripts in `packages\backend\package.json`)

I'm not completely sure about this approach, but I've crafted it to have as an alternative to what we've got so far. This is open for discussion.

---

NOTE: 

- `ensureAndLoadEnv` is placed in `@dddforum/shared` (not in `@dddforum/cli`) because `@dddforum/cli` is intended to be used as a dev dependency only, but we need `ensureAndLoadEnv` to use in the production code as well, e.g. in `packages\backend\src\index.ts`.
- `jest` by default sets `NODE_ENV` to `test`. It is possible to override, but I simply added `.env.test` files so far.
- @donnyroufs suggested using `tsup`, so we might don't need `ts-alias`. I will evaluate it a little bit later. UPD. evaluated in https://github.com/stemmlerjs/ddd-forum-bp-first/pull/7#issuecomment-1637001232.

---

This is already integrated into the NPM scripts. Just try running:

- `npm run test:e2e` // run tests in test (dev) env
- `npm run start:dev:backend` // prepare a dev env and start the backend in dev mode
- `npm run build` // builds everything
- `npm run start:ci:backend` // start backend within the staging env
etc.

All commands print output that provides info on what evn / tsconfig / package.json is used. You can add more scripts to satisfy different needs. What env is used is controlled by `NODE_ENV=<env>` in the scripts.

### CODE REVIEW

File: packages/buildTools/cli/cli.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+import { bindMigrate } from './bindMigrate';
+import { bindPrepareEnv } from './bindPrepareEnv';
+
+const program = new Command();
+
+program.name('@DDD forum build utils').description('CLI to build @DDD forum packages').version(version);
+
+bindBuild(program);
+bindPrepareEnv(program);
+bindMigrate(program);
```
Comment (stemmlerjs) [inline packages/buildTools/cli/cli.ts:16]: That's sweet. Very open-closed. We could add new scripts to this as we like. That's awesome. 

Is it right to call this a _build_ tool? Seems like we can use it for most of what we have to do.

Here's my current list of scripts/capabilities we will need from what I gather.

Screen Shot 2023-07-15 at 8 01 34 PM

Comment (karpov-kir) [reply packages/buildTools/cli/cli.ts:16]: @stemmlerjs,

> Is it right to call this a build tool?

Suggest your name. 🙂 How about just `cli`?

Comment (stemmlerjs) [reply packages/buildTools/cli/cli.ts:16]: @karpov-kir cli is perfect. I wrote some more notes on this and what I expect students to be able to do for the first E2E on the backend assignment. https://khalilstemmler.notion.site/Action-E2E-test-on-the-Backend-418f445837c4470b8bc5a1c1449960d4?pvs=4

**Comment by karpov-kir** [comment]
@stemmlerjs, regarding `tsup`.

This is a faster code compiler that simply compiles TS files to JS files without type-checking https://tsup.egoist.dev/#what-about-type-checking.

image

It provided better performance, but I haven't found any way to use this tool to replace paths like "@dddforum/shared/src/..." to use "dist" instead.

I wouldn't use this tool at least for now because:

- It is a little bit less explicit than just `tsc`
  - It does not respect `tsconfig.json` and it has its own configuration
- We still would need to use `ts-alias` (or configure funky TS aliases)
  - With this tool, we would still need to do `tsup ... && tsc-alias ...`
- The DDD forum project is not that big to use more performant tools instead of `tsc`

File: tsconfig.json [hunk 2]
```diff
@@ -1,17 +1,12 @@
...
-    "lib": ["es6"],
+    "target": "ES6",
+    "module": "CommonJS",
+    "moduleResolution": "Node",
     "strict": true,
     "esModuleInterop": true,
     "types": ["node", "jest"],
-    "resolveJsonModule": true,
-    "paths": {
-      "@dddforum/backend/*": ["./packages/backend/*"],
```
Comment (karpov-kir) [inline tsconfig.json:12]: @stemmlerjs, when we use `src` imports we actually don't need these paths then.

File: packages/shared/README.md [hunk 3]
```diff
@@ -2,4 +2,4 @@
 
 This is the shared folder.
 
-Any files you want to share in here across both frontend and backend MUST be have the .shared extension.
```
Comment (karpov-kir) [inline packages/shared/README.md:5]: @stemmlerjs, why is this `.shared` rule? Should we also then prefix `packages\shared\src\ensureAndLoadEnv.ts`, `packages\shared\src\logger.ts`, etc.?

File: packages/shared/README.md [hunk 4]
```diff
@@ -2,4 +2,4 @@
...
\ No newline at end of file
+Any files you want to share in here across both frontend and backend MUST be have the .shared extension. This is intended to be used as normal (not dev) dependency in other packages.
```
Comment (karpov-kir) [inline packages/shared/README.md:5]: @stemmlerjs, why is this `.shared` rule? Should we also then prefix `packages\shared\src\ensureAndLoadEnv.ts`, `packages\shared\src\logger.ts`, etc.?

File: packages/buildTools/cli/cli.ts (hunk 1 above)
Comment (karpov-kir) [reply packages/buildTools/cli/cli.ts]: @stemmlerjs, done in https://github.com/stemmlerjs/ddd-forum-bp-first/pull/7/commits/60cd9664d7f1e1793a2d567ce6a68b51408c7248.

**Comment by stemmlerjs** [comment]
## Problem

You need to run scripts to develop, test, build, and deploy software but doing so gets kind of messy, and not everyone is very good at writing scripts.

### Agitate

1. Very messy npm scripts calling each other
2. Impossible to reason what you _can do_ and _how things are done_ because of the messiness
3. No useful errors and safeguards when running npm scripts: Scripts without constraints and rules which don't give you clarity on how to use them correctly. Lots of guesswork leads to low DX and others have to run into the same problems.
4. Hard to manage switching and controlling current env: You need to manage to do this in different environments, and it's cognitively to keep track of and figure out how to manage environment variables
5. Can't test scripts without specific env: When your scripts are messy and rely on a specific environment (such as the CI), you end up having to run everything in the CI and this can take a long time for you to get things working, because you have to wait for it to finish to test your changes (lack of control over testing; can't test scripts properly)

Ultimately, this all leads to confusion and feeling overwhelmed with your scripts and lacking clarity on how to 

## Solution

You want a tool that 1) you own, 2) gets better over time because you add in constraints, rules, new functionality, 3) gives you control of the environment, 4) highly configurable, 5) common challenges - like ensuring env vars are loaded that need to be loaded are handled, frontend and backend 6) is simple and has high DX (discoverability & understanding)

## How to use the tool

### Examples

> Aside: We should always drive everything we do through solid, concrete examples. Never just through building to build. Always have a valuable scenario in mind to implement new use cases and functionality.

1. Ensure that when I build my frontend and backend in production mode, I can verify that it can access the correct env vars, and I know where they're getting loaded from.

2. I want to configure my tool to rebuild the Prisma schema before starting the backend; I can pass in a config or extend the tool to do this easily.

3. I want to configure a command or sequence of commands to get executed on every single command, with the ability to exclude which commands it does not run for. 

  - Build, start on backend and frontend
- How to extend the tool:

**Comment by karpov-kir** [comment]
I'm closing this PR for now since the focus has changed to target other important things and let's revisit it if we need a CLI package again.

**Pull request closed without merging**